### PR TITLE
Request Telemetry: Collecting Uri from HttpRequest

### PR DESCRIPTION
**Pull request by Sergei-Nikitin**

### CODE REVIEW

File: src/Microsoft.ApplicationInsights.AspNet/RequestTrackingMiddleware.cs [hunk 1]
```diff
@@ -50,6 +52,7 @@ public RequestTrackingMiddleware(RequestDelegate next, TelemetryClient client)
                     telemetry.ResponseCode = httpContext.Response.StatusCode.ToString();
                     telemetry.Success = httpContext.Response.StatusCode < 400;
                     telemetry.HttpMethod = httpContext.Request.Method;
+                    telemetry.Url = httpContext.Request.GetUri();
```
Comment (SergKanz) [inline src/Microsoft.ApplicationInsights.AspNet/RequestTrackingMiddleware.cs]: since this method can throw we need to create a try/catch statement

**Comment by SergKanz** [comment]
:+1:

File: test/runAllTests.cmd [hunk 2]
```diff
@@ -1,8 +1,11 @@
-%USERPROFILE%\.dnx\runtimes\dnx-clr-win-x86.1.0.0-beta4\bin\dnx.exe .\Microsoft.ApplicationInsights.AspNet.Tests test
-%USERPROFILE%\.dnx\runtimes\dnx-coreclr-win-x86.1.0.0-beta4\bin\dnx.exe .\Microsoft.ApplicationInsights.AspNet.Tests test
+@echo off
 
-%USERPROFILE%\.dnx\runtimes\dnx-clr-win-x86.1.0.0-beta4\bin\dnx.exe .\Mvc6Framework45.FunctionalTests test
-%USERPROFILE%\.dnx\runtimes\dnx-coreclr-win-x86.1.0.0-beta4\bin\dnx.exe .\Mvc6Framework45.FunctionalTests test
+FOR %%A IN (dnx-clr-win-x86.1.0.0-beta4 dnx-coreclr-win-x86.1.0.0-beta4) DO (
```
Comment (olegsych) [inline test/runAllTests.cmd]: The force is strong with this one. :grinning: 
I can never remember how to do loops in batch script, which is why I asked about PowerShell in @abaranch pull request.

File: test/Microsoft.ApplicationInsights.AspNet.Tests/project.json [hunk 3]
```diff
@@ -8,10 +8,11 @@
         "Microsoft.AspNet.Mvc.Core": "6.0.0-beta4-*",
         "Microsoft.Framework.Logging": "1.0.0-beta4-*",        
         "Microsoft.Framework.ConfigurationModel.Json": "1.0.0-beta4-*",        
-        "xunit.runner.aspnet": "2.0.0-aspnet-beta4-*"
+        "xunit": "2.1.0-beta2-*",
+        "xunit.runner.dnx": "2.1.0-beta2-*"
```
Comment (olegsych) [inline test/Microsoft.ApplicationInsights.AspNet.Tests/project.json]: Is this a replacement for `xunit.runner.aspnet`? Does it work in VS 2015 RC as well?

File: src/Microsoft.ApplicationInsights.AspNet/Extensions/HttpRequestExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            {
+                throw new ArgumentException("Http request Scheme is not specified");
+            }
+
+            if (false == request.Host.HasValue)
+            {
+                throw new ArgumentException("Http request Host is not specified");
+            }
+
+            var builder = new StringBuilder();
```
Comment (olegsych) [inline src/Microsoft.ApplicationInsights.AspNet/Extensions/HttpRequestExtensions.cs]: Can we use the [UriBuilder](https://msdn.microsoft.com/en-us/library/system.uribuilder.aspx) here?

**Comment by abaranch** [comment]
Approved

File: src/Microsoft.ApplicationInsights.AspNet/Extensions/HttpRequestExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
+﻿namespace Microsoft.ApplicationInsights.AspNet.Extensions
+{
+    using Microsoft.AspNet.Http;
+    using System;
+    using System.Text;
+
+    /// <summary>
+    /// Set of extension methods for Microsoft.AspNet.Http.HttpRequest
+    /// </summary>
+    public static class HttpRequestExtensions
```
Comment (olegsych) [inline src/Microsoft.ApplicationInsights.AspNet/Extensions/HttpRequestExtensions.cs]: `HttpRequestExtensions` is used only by the `RequestTrackingMiddleware` correct? Why do we want it to be public? 
This imposes the requirement to test arguments and throw exceptions here, leading to ask by @SergKanz to catch them in the `RequestTrackingMiddleware`.

**Comment by upendras** [comment]
:shipit:

File: test/Microsoft.ApplicationInsights.AspNet.Tests/RequestTrackingMiddlewareTest.cs [hunk 6]
```diff
@@ -8,24 +8,64 @@
...
+        public RequestTrackingMiddlewareTest()
+        {
+            this.middleware = new RequestTrackingMiddleware(
+                this.nextMiddleware,
+                CommonMocks.MockTelemetryClient(telemetry => this.sentTelemetry = telemetry));
+        }
+
         [Fact]
-        public async Task SdkVersionIsPopulatedByMiddleware()
+        public async Task TestSdkVersionIsPopulatedByMiddleware()
```
Comment (olegsych) [inline test/Microsoft.ApplicationInsights.AspNet.Tests/RequestTrackingMiddlewareTest.cs]: Definitely a nit pick: I tend to find the term `Test` in a test method name wasteful. It is already implied by the `Fact` or `TestMethod` attribute and only takes up space on the editor screen and in test explorer.

File: src/Microsoft.ApplicationInsights.AspNet/RequestTrackingMiddleware.cs (hunk 1 above)
Comment (olegsych) [inline src/Microsoft.ApplicationInsights.AspNet/RequestTrackingMiddleware.cs]: @Sergei-Nikitin and I respectfully disagree :smile: and will try to convince you later tonight :facepunch:

**Pull request merged**